### PR TITLE
feat(codec/scale): support enum and enum class

### PR DESCRIPTION
**Pull request by xedonman**

### CODE REVIEW

**Comment by xedonman** [comment]
@conr2d could you check encoding/decoding of c/c++ enum types 
enumeration encoding of SCALE which is described below
https://github.com/paritytech/parity-scale-codec#simple-types

but I just encoded into underlying type of enum because features of enum/ enum class is different from that of rust

and also is there a naming convention for concept?
I just temporarily named it enum_c which will be renamed

File: src/noir/codec/scale.h [hunk 1]
```diff
@@ -47,6 +47,21 @@ datastream<Stream>& operator>>(datastream<Stream>& ds, uint256_t& v) {
   return ds;
 }
 
+template<typename Stream, enum_c E>
+datastream<Stream>& operator<<(datastream<Stream>& ds, const E& v) {
+  auto data = static_cast<const std::underlying_type_t<E>>(v);
+  ds << data;
```
Comment (conr2d) [inline src/noir/codec/scale.h]:
```suggestion
  ds << static_cast<const std::underlying_type_t<E>>(v);
```

File: src/noir/codec/scale.h [hunk 2]
```diff
@@ -47,6 +47,21 @@ datastream<Stream>& operator>>(datastream<Stream>& ds, uint256_t& v) {
...
+  return ds;
+}
+
+template<typename Stream, enum_c E>
+datastream<Stream>& operator>>(datastream<Stream>& ds, E& v) {
+  auto data = static_cast<std::underlying_type_t<E>>(v);
+  ds >> data;
+  v = static_cast<E>(data);
```
Comment (conr2d) [inline src/noir/codec/scale.h]:
```suggestion
  std::underlying_type_t<E> tmp;
  ds >> tmp;
  v = static_cast<E>(tmp);
```

File: src/noir/codec/scale.h [hunk 3]
```diff
@@ -47,6 +47,21 @@ datastream<Stream>& operator>>(datastream<Stream>& ds, uint256_t& v) {
   return ds;
 }
 
+template<typename Stream, enum_c E>
```
Comment (conr2d) [inline src/noir/codec/scale.h]: There is no naming convention related to concepts, but what about `noir::enumeration` for this?

```suggestion
template<typename Stream, enumeration E>
```

File: src/noir/common/concepts.h [hunk 4]
```diff
@@ -11,6 +11,9 @@ namespace noir {
 template<typename T>
 concept integral = std::is_integral_v<T>;
 
+template<typename E>
+concept enum_c = std::is_enum_v<E>;
```
Comment (conr2d) [inline src/noir/common/concepts.h]:
```suggestion
concept enumeration = std::is_enum_v<E>;
```

**Comment by xedonman** [comment]
pushed suggested changes

**Review by conr2d — APPROVED** [review]

**Pull request merged**